platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java

```diff
@@ -1089,24 +1098,56 @@ public void removeAllMarkers() {
      */
     public void setMapChangeListener(@Nullable final MapChangeListener listener) {
         mapChangeListener = listener;
+        mapRegionChangeStateManager = new MapRegionChangeStateManager(mapChangeListener);
     }
 
-    //Convenience member functions
-    private void onRegionWillChange(boolean animated) {
-        if (mapChangeListener != null) {
-            mapChangeListener.onRegionWillChange(animated);
+    private static class MapRegionChangeStateManager {
+
+        enum MapRegionChangeState {
+            IDLE,
+            JUMPING,
+            ANIMATING,
         }
-    }
 
-    private void onRegionDidChange(boolean animated) {
-        if (mapChangeListener != null) {
-            mapChangeListener.onRegionDidChange(animated);
+        WeakReference<MapChangeListener> mapChangeListenerWeakReference;
+
+        MapRegionChangeStateManager(@Nullable MapChangeListener listener) {
+            mapChangeListenerWeakReference = new WeakReference<>(listener);
         }
-    }
 
-    private void onRegionIsChanging() {
-        if (mapChangeListener != null) {
-            mapChangeListener.onRegionIsChanging();
+        MapRegionChangeState currentState = MapRegionChangeState.IDLE;
+
+        void setMapRegionState(MapRegionChangeState state) {
+
+            MapChangeListener mapChangeListener = mapChangeListenerWeakReference.get();
+            if (mapChangeListener != null) {
+                switch (currentState) {
+                    case IDLE:
+                        if (state == MapRegionChangeState.JUMPING) {
+                            mapChangeListener.onRegionWillChange(false);
+                        } else if (state == MapRegionChangeState.ANIMATING){
+                            mapChangeListener.onRegionWillChange(true);
+                        }
+                        break;
+                    case JUMPING:
+                        if (state == MapRegionChangeState.IDLE) {
+                            mapChangeListener.onRegionDidChange(false);
+                        } else if (state == MapRegionChangeState.ANIMATING) { // e.g. panning to fling
+                            mapChangeListener.onRegionIsChanging();
+                        }
+                        break;
+                    case ANIMATING:
+                        if (state == MapRegionChangeState.IDLE) {
+                            mapChangeListener.onRegionDidChange(true);
+                        } else if (state == MapRegionChangeState.ANIMATING) {
+                            mapChangeListener.onRegionIsChanging();
+                        } else { // ?? ANIMATING -> JUMPING?? Possible?
```

Comment:
I'd prefer `else if (state == MapRegionChangeState.JUMPING)` just to be clearer, and in case we add more states.

Reply:
I don't _think_ this transition is ever triggered - animation should be canceled and set the state to `IDLE` before it becomes `JUMPING`. I guess we could find out for sure by putting a breakpoint here or throwing an exception.

Reply:
Yeah I was not able to trigger this at all. Wanted to keep it in there for discussion point of view.

Reply:
Going to remove this. Do not think it makes sense to respect this state transition of `ANIMATING`->`JUMPING`. The transitions of `DID_ANIMATING`, `WILL_CHANGE` implying an `IDLE` state in between will always happen.